### PR TITLE
Implement EXPAND syntax

### PR DESCRIPTION
**Pull request by pengzhout**
Implement EXPAND syntax

    Implement "ALTER TABLE table EXPAND TABLE" to expand tables.

    "Expanding" and "Set Distributed by" are actually two different kind of
    operations on tables, old gpexpand used to use "Set Distributed by" to
    expand tables for historical reasons and our early version of expand
    were also squashed into "Set Distributed by", this make code hard to
    hard to understand and concept confused.

    This commit divide "Expanding" and "Set Distributed by" totally and
    implement "Expanding" with new syntax. We have two method to implement
    data movement, one is CTAS, another is RESHUFFLE, depend on how much
    data need to move. If tuples to move is less than 10000, choose
    RESHUFFLE, or if scale to move is less than 30% choose RESHUFFLE,
    otherwise, choose CTAS

    The new syntax is supposed to only used by gpexpand and not be exposed
    to normal users, so no need to update document.

For partition table, we disallow expand leaf partition seperately because
root partition cannot has different numsegments with leaf partitions,
SELECT/UPDATE should be fine if numsegments is inconsistent, however,
INSERT will make trouble that data are inserted to unexpected place.

### CODE REVIEW

**Comment by pengzhout** [comment]
This PR is separated from PR #6289, it contains the part that has been accepted by reviewers, I want to merge this part as soon as possible.

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -14423,8 +14475,8 @@ ReshuffleRelationData(Relation rel)
 	{
 		ereport(NOTICE,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("Do not need to reshuffle %s.%s",
-							   namespace_name,prelname)));
+				 errmsg("Do not need to reshuffle %s.%s",
+						namespace_name,prelname)));
```
Comment (kainwen) [inline src/backend/commands/tablecmds.c]: add a space `namespace_name,prelname` .

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -14521,6 +14573,374 @@ ReshuffleRelationData(Relation rel)
...
+	Oid					relid = RelationGetRelid(rel);
+	GpPolicy			*newPolicy;
+	GpPolicy			*policy = rel->rd_cdbpolicy;
+
+	if (Gp_role == GP_ROLE_UTILITY)
+		ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("EXPAND not supported in utility mode")));
+
+	/* Permissions checks */
```
Comment (kainwen) [inline src/backend/commands/tablecmds.c]: Can we move the two checks in ATPrepCmd?

**Comment by pengzhout** [comment]
rebased with master, ready for merge if pr pipeline pass

**Comment by pengzhout** [comment]
merged, thanks for the review.

**Comment by hlinnaka** [comment]
On 23/11/2018 05:11, Zhenghua Lyu wrote:
> In src/backend/commands/tablecmds.c 
> <https://github.com/greenplum-db/gpdb/pull/6301#discussion_r235839432>:
> 
>> @@ -14423,8 +14475,8 @@ ReshuffleRelationData(Relation rel)
>   	{
>   		ereport(NOTICE,
>   				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
> -						errmsg("Do not need to reshuffle %s.%s",
> -							   namespace_name,prelname)));
> +				 errmsg("Do not need to reshuffle %s.%s",
> +						namespace_name,prelname)));
> 
> add a space |namespace_name,prelname| .

And I still think we should not throw an error if you ask for a no-op 
ALTER TABLE. If you think EXPAND should work differently from all other 
ALTER TABLE variants, please explain why.

- Heikki

**Comment by hlinnaka** [comment]
I think it was premature to push this. It had been open for less than 24 hours, and I had raised comments on the behaviour of EXPAND TABLE on a table that doesn't need expanding (no-op, or throw an error?). And I thought the syntax was still under discussion. Have you thought about shrinking yet?

I had not even started to look at the details here, just the high level stuff. In the form that this was committed, all the PartStatus stuff is unused. There's a lot of duplicated code between ATExecSetDistributedBy() and ATExecExpandTableCTAS(). Could we use the reshuffle method to also implement SET DISTRIBUTED BY? Would that make sense?

The heuristics on which method to use seem quite arbitary. I don't have any better ideas there, but we should also consider Scott Kahler's worry about running out of disk space, if the CTAS method is used. In the reshuffle method, would it be possible to rebuild the indexes after moving all the data, instead of updating them in retail?

**Comment by hlinnaka** [comment]
I opened PR https://github.com/greenplum-db/gpdb/pull/6308 to clean this up a little bit. But it doesn't address all my concerns. Please review, and please look at the other things I mentioned, next.

**Comment by pengzhout** [comment]
> I think it was premature to push this. It had been open for less than 24 hours, and I had raised comments on the behaviour of EXPAND TABLE on a table that doesn't need expanding (no-op, or throw an error?). And I thought the syntax was still under discussion. Have you thought about shrinking yet?

I thought we had a conclusion that the first part is ok and had been accepted, so did a quick push so we can do a further cleanup of ATExecSetDistributedBy() and can draft gpexpand PR depends on it.

> The heuristics on which method to use seem quite arbitary. I don't have any better ideas there, but we should also consider Scott Kahler's worry about running out of disk space, if the CTAS method is used. In the reshuffle method, would it be possible to rebuild the indexes after moving all the data, instead of updating them in retail?

Indeed, we have not experimental rules to decide the method yet, so just implemented the simplest logic in the first version without considering of index and table size or if it's append-only table or not. we need some tests and more discussion to finalize this logical.

**Pull request merged**